0000-dom-query-testing-interface.md

```diff
@@ -0,0 +1,265 @@
+
+As a case study (and most proximate motivation for this RFC), let's consider `fractal-page-object`. Currently `fractal-page-object` integrates with `@ember/test-helpers` and `qunit-dom` by exposing an `element` property on all page objects, and that works because their APIs accept direct `Element` references. But this has three major drawbacks:
+
+1. It only supports single elements. While this isn't an issue with the `@ember/test-helpers` DOM APIs, in `qunit-dom` that means there's no way to use the multi-element assertions, analogous to  `assert.dom(document.querySelectorAll('.list-item')).exists({ count: 3 })`.
+2. It does not produce useful information when encountering an error/failure. For example, `await click('.does-not-exist')` errors with `Error: Element not found when calling click('.does-not-exist').`, while `await click(document.querySelector('.does-not-exist'))` errors with the less helpful `Error: Must pass an element or selector to 'click'` (because it doesn't have any other information to show -- all it knows is it was passed `null` instead of a selector or `Element`). Similarly, `assert.dom('.does-not-exist').exists()` produces the assertion message `Element .does-not-exist exists`, while `assert.dom(document.querySelector('.does-not-exist')).exists()` produces the less helpful `Element <not found> exists`.
+3. It supports a noticeably less ergonomic integration with page objects. In `fractal-page-object`, the integrations currently look like `await click(page.listItems[2].checkbox.element)` and `assert.dom(page.listItems[2].checkbox.element).isChecked()`, rather than `await click(page.listItems[2].checkbox)` and `assert.dom(page.listItems[2].checkbox).isChecked()`. While it may look like a minor difference, in practice it's quite a stumbling block as the mental model of page objects encourages users to think of page object nodes (e.g. `page.listItems[2].checkbox`) as a proxy for the DOM element(s) they describe, so forgetting to add the `.element` is extremely common, even for the author of the library!
+
+We can solve these by formalizing the notion of DOM element descriptors in an interface, so page object implementations and any other mechanisms for DOM access can produce objects implementing this interface and they can be passed to these test helpers without incurring any of the shortcoming described above.
+
+## Detailed design
+
+### The interfaces
+
+This RFC proposes implementing a very lightweight library that exports one `Symbol` and two interfaces:
+
+```ts
+export const ELEMENT_DESCRIPTOR = Symbol('element descriptor');
+
+export interface IElementDescriptor {
+  element: Element | null;
+  elements: NodeListOf<Element> | Iterable<Element>;
+  description?: string;
+}
+
+export interface IElementDescriptorProvider {
+  [ELEMENT_DESCRIPTOR]: IElementDescriptor;
+}
```

Comment:
We discussed this generally in discord chat, and I think I'd like to go a slightly different direction:

```ts
function retrieveElementDescriptor<T extends object>(obj: T): IElementDescriptor;
function associateElementDescriptor<T extend object>(obj: T, retriever: retrieveElementDescriptor<T>);
```

This allows us to use (essentially) a single system and fits in with the other APIs that we have (e.g. `associateDestroyableChild`).